kie-server-parent/kie-server-services/kie-server-services-optaplanner/src/main/java/org/kie/server/services/optaplanner/SolverServiceBase.java

```diff
@@ -0,0 +1,323 @@
+                    null );
+        }
+    }
+
+    public ServiceResponse<SolverInstance> createSolver(String containerId, String solverId, SolverInstance instance) {
+        if ( instance == null || instance.getSolverConfigFile() == null ) {
+            logger.error( "Error creating solver. Configuration file name is null: " + instance );
+            return new ServiceResponse<SolverInstance>(
+                    ServiceResponse.ResponseType.FAILURE, "Failed to create solver for container " + containerId +
+                                                          ". Solver configuration file is null: " + instance );
+        }
+        instance.setContainerId( containerId );
+        instance.setSolverId( solverId );
+
+        try {
+            KieContainerInstanceImpl ci = context.getContainer( containerId );
+            if ( ci == null ) {
+                logger.error( "Error creating solver. Container does not exist: " + containerId );
+                return new ServiceResponse<SolverInstance>( ServiceResponse.ResponseType.FAILURE, "Failed to create solver. Container does not exist: " + containerId );
+            }
+
+            // have to synchronize on the ci or a concurrent call to dispose may create inconsistencies
+            synchronized ( ci ) {
+                if( solvers.containsKey( instance.getSolverInstanceKey() ) ) {
+                    logger.error( "Error creating solver. Solver '" + solverId + "' already exists for container '" + containerId + "'." );
+                    return new ServiceResponse<SolverInstance>( ServiceResponse.ResponseType.FAILURE, "Failed to create solver. Solver '" + solverId +
+                                                                                                      "' already exists for container '" + containerId + "'." );
+                }
+                SolverInstanceContext sic = new SolverInstanceContext( instance );
+
+                try {
+                    SolverFactory<?> solverFactory = SolverFactory.createFromKieContainerXmlResource( ci.getKieContainer(), instance.getSolverConfigFile() );
+
+                    Solver<?> solver = solverFactory.buildSolver();
+
+                    sic.setSolver( solver );
+                    updateSolverInstance( sic );
+
+                    solvers.put( instance.getSolverInstanceKey(), sic );
+
+                    logger.info( "Solver '" + solverId + "' successfully created in container '" + containerId + "'" );
+                    return new ServiceResponse<SolverInstance>(ServiceResponse.ResponseType.SUCCESS,
+                                                               "Solver '" + solverId + "' successfully created in container '" + containerId + "'",
+                                                               instance );
+
+                } catch( Exception e ) {
+                    logger.error("Error creating solver factory for solver " + instance, e);
```

Comment:
Why duplicate this exception handling inside and outside the synchronized statement? I don't see any race condition if it was simply all handled by the outer exception handling.